File: knapsack/knapsack.cpp
//********* knapsack.cpp *********

#include <iostream>
#include <fstream>
#include <vector>
#include "localsolver.h"

using namespace localsolver;
using namespace std;

class Knapsack {
public:
    // Number of items. 
    int nbItems;

    // Items properties. 
    vector<lsint> weights;
    vector<lsint> values;

    // Knapsack bound 
    lsint knapsackBound;
    
    // LocalSolver. 
    LocalSolver localsolver;

    // Decision variables. 
    vector<LSExpression> x;

    // Objective. 
    LSExpression knapsackValue;

    // Solution (items in the knapsack). 
    vector<int> solution;

    // Reads instance data. 
    void readInstance(const string& fileName) {
        ifstream infile;
        infile.exceptions(ifstream::failbit | ifstream::badbit);
        infile.open(fileName.c_str());

        infile >> nbItems;

        weights.resize(nbItems);
        for (int i = 0; i < nbItems; i++)
            infile >> weights[i];

        values.resize(nbItems);
        for (int i = 0; i < nbItems; i++)
            infile >> values[i];
    
        infile >> knapsackBound;   
    }

    void solve(int limit) {
        // Declares the optimization model. 
        LSModel model = localsolver.getModel();

        // Decision variables x[i] 
        x.resize(nbItems);
        for (int i = 0; i < nbItems; i++) {
            x[i] = model.boolVar();
        }

        // Weight constraint
        LSExpression knapsackWeight = model.sum();
        for (int i = 0; i < nbItems; i++) {
            LSExpression itemWeight = x[i]*weights[i];
            knapsackWeight += itemWeight;
        }
        model.constraint(knapsackWeight <= knapsackBound);
    
        // Maximize value
        knapsackValue = model.sum();
        for (int i = 0; i < nbItems; i++) {
            LSExpression itemValue = x[i]*values[i];
            knapsackValue += itemValue;
        }
        model.maximize(knapsackValue);
        model.close();

        // Parameterizes the solver. 
        localsolver.getParam().setTimeLimit(limit);

        localsolver.solve();

        solution.clear();
        for (int i = 0; i < nbItems; ++i)
            if (x[i].getValue() == 1) 
                solution.push_back(i);
    }

    // Writes the solution in a file 
    void writeSolution(const string& fileName) {
        ofstream outfile;
        outfile.exceptions(ofstream::failbit | ofstream::badbit);
        outfile.open(fileName.c_str());

        outfile << knapsackValue.getValue() << endl;
        for (unsigned int i = 0; i < solution.size(); ++i)
            outfile << solution[i] << " ";
        outfile << endl;
    }
};

int main(int argc, char** argv) {
    if (argc < 2) {
        cerr << "Usage: knapsack inputFile [outputFile] [timeLimit]" << endl;
        return 1;
    }

    const char* instanceFile = argv[1];
    const char* solFile = argc > 2 ? argv[2] : NULL;
    const char* strTimeLimit = argc > 3 ? argv[3] : "20";

    try {
        Knapsack model;
        model.readInstance(instanceFile);
        model.solve(atoi(strTimeLimit));
        if (solFile != NULL) model.writeSolution(solFile);
        return 0;
    } catch (const exception& e) {
        cerr << "An error occurred: " << e.what() << endl;
        return 1;
    }
}

File: knapsack/Knapsack.java
/********** Knapsack.java **********/

import java.util.*;
import java.io.*;
import localsolver.*;

public class Knapsack {
    // Number of items. 
    private int nbItems;

    // Items properties. 
    private int[] weights;
    private int[] values;

    // Knapsack bound 
    private int knapsackBound;

    // Solver. 
    private final LocalSolver localsolver;

    // LS Program variables. 
    private LSExpression[] x;

    // Objective. 
    private LSExpression knapsackValue;
    
    // Solutions (classes at each position). 
    private List<Integer> solutions;

    private Knapsack(LocalSolver localsolver) {
        this.localsolver = localsolver;
    }

    // Reads instance data. 
    private void readInstance(String fileName) throws IOException {
        try (Scanner input = new Scanner(new File(fileName))) {
            nbItems = input.nextInt();

            weights = new int[nbItems];
            for (int i = 0; i < nbItems; i++) {
                weights[i] = input.nextInt();
            }

            values = new int[nbItems];
            for (int i = 0; i < nbItems; i++) {
                values[i] = input.nextInt();
            }
        
            knapsackBound = input.nextInt();
        }
    }

    private void solve(int limit) {
        // Declares the optimization model. 
        LSModel model = localsolver.getModel();

        // boolean variables x[i]
        x = new LSExpression[nbItems];
        for (int i = 0; i < nbItems; i++) {
            x[i] = model.boolVar();
        }

        // weight constraint
        LSExpression knapsackWeight = model.sum();
        for (int i = 0; i < nbItems; i++) {
            LSExpression itemWeight = model.prod(x[i], weights[i]);
            knapsackWeight.addOperand(itemWeight);
        }        
        model.constraint(model.leq(knapsackWeight, knapsackBound));

        // maximize value
        knapsackValue = model.sum();
        for (int i = 0; i < nbItems; i++) {
            LSExpression itemValue = model.prod(x[i], values[i]);
            knapsackValue.addOperand(itemValue);
        }
    
        model.maximize(knapsackValue);
        model.close();

        // Parameterizes the solver. 
        localsolver.getParam().setTimeLimit(limit);

        localsolver.solve();

        solutions = new ArrayList<Integer>();
        for (int i = 0; i < nbItems; i++) 
            if (x[i].getValue() == 1) 
        solutions.add(i);         
    }

    // Writes the solution in a file 
    private void writeSolution(String fileName) throws IOException {
        try (PrintWriter output = new PrintWriter(fileName)) {
            output.println(knapsackValue.getValue());
            for (int i = 0; i < solutions.size(); ++i) 
                output.print(solutions.get(i) + " ");
            output.println();
        }
    }
    
     public static void main(String[] args) {
        if (args.length < 1) {
            System.err.println("Usage: java Knapsack inputFile [outputFile] [timeLimit]");
            System.exit(1);
        }

        String instanceFile = args[0];
        String outputFile = args.length > 1 ? args[1] : null;
        String strTimeLimit = args.length > 2 ? args[2] : "20";

         try (LocalSolver localsolver = new LocalSolver()) {
            Knapsack model = new Knapsack(localsolver);
            model.readInstance(instanceFile);
            model.solve(Integer.parseInt(strTimeLimit));
            if (outputFile != null) {
                model.writeSolution(outputFile);
            }
        } catch(Exception ex) {
            System.err.println(ex);
            ex.printStackTrace();
            System.exit(1);
        }
    }
}
